src/algorithms/douglas_rachford.jl

```diff
@@ -41,10 +41,10 @@ Base.IteratorSize(::Type{<:DouglasRachfordIteration}) = Base.IsInfinite()
 
 Base.@kwdef struct DouglasRachfordState{Tx}
     x::Tx
-    y::Tx = zero(x)
-    r::Tx = zero(x)
-    z::Tx = zero(x)
-    res::Tx = zero(x)
+    y::Tx = similar(x)
+    r::Tx = similar(x)
+    z::Tx = similar(x)
+    res::Tx = similar(x)
```

Comment:
Is this completely safe? Sometimes similar returns NaNs

Reply:
Should be safe, as long as the array is initialized before it's used, otherwise there's a problem in Julia `Base`!

The thing is that `similar` is faster than `zero` or `copy` (I can get numbers about how faster). Actually, I should check if there's other places where `similar` could be used!

Reply:
check this out:
```julia
julia> x = zeros(10); any([any(isnan.(similar(x))) for i = 1:1000])
true
```

Reply:
I just hate bugs involving 'NaN's. 😆  Is there really a significant speed up?

Reply:
I’ll check, I’m just obsessing a bit over getting rid of any superfluous operation 😅

Reply:
For `n = 1000` and using `Float64`, `similar` is about 4x faster than `zero`. The time ratio gets larger as `n` increases (which makes sense)

```
julia> n = 1000; x = randn(n);

julia> @benchmark zero($x)
BenchmarkTools.Trial: 10000 samples with 198 evaluations.
 Range (min … max):  424.631 ns … 42.357 μs  ┊ GC (min … max):  0.00% … 97.30%
 Time  (median):       1.062 μs              ┊ GC (median):     0.00%
 Time  (mean ± σ):     2.132 μs ±  5.725 μs  ┊ GC (mean ± σ):  50.82% ± 17.80%

  ▅█                                                           ▁
  ██▇▅▅▇▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▃▃▄▄▅▅▅▅▄▅▅▅▆▆▇█▇▇▆▇▇ █
  425 ns        Histogram: log(frequency) by time      33.9 μs <

 Memory estimate: 7.94 KiB, allocs estimate: 1.

julia> @benchmark similar($x)
BenchmarkTools.Trial: 5065 samples with 883 evaluations.
 Range (min … max):  155.805 ns … 10.300 μs  ┊ GC (min … max):  0.00% … 96.60%
 Time  (median):     283.980 ns              ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.112 μs ±  1.958 μs  ┊ GC (mean ± σ):  74.49% ± 34.68%

  █▇▁                                ▂▂▁▁▁                     ▁
  ███▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▁▆▆▆▄▄████████▇▇▇█▇█▇██▇▇█▇▇▇▇▇▆ █
  156 ns        Histogram: log(frequency) by time      7.48 μs <

 Memory estimate: 7.94 KiB, allocs estimate: 1.
```